File: backend/internal/types/database.types.go
package types

type PublicProfilesSelect struct {
	CreatedAt     string  `json:"created_at"`
	CreditBalance int32   `json:"credit_balance"`
	Email         string  `json:"email"`
	Id            string  `json:"id"`
	Role          string  `json:"role"`
	UpdatedAt     *string `json:"updated_at"`
	Username      string  `json:"username"`
}

type PublicProfilesInsert struct {
	CreatedAt     *string `json:"created_at"`
	CreditBalance *int32  `json:"credit_balance"`
	Email         string  `json:"email"`
	Id            string  `json:"id"`
	Role          *string `json:"role"`
	UpdatedAt     *string `json:"updated_at"`
	Username      string  `json:"username"`
}

type PublicProfilesUpdate struct {
	CreatedAt     *string `json:"created_at"`
	CreditBalance *int32  `json:"credit_balance"`
	Email         *string `json:"email"`
	Id            *string `json:"id"`
	Role          *string `json:"role"`
	UpdatedAt     *string `json:"updated_at"`
	Username      *string `json:"username"`
}

type PublicEquipmentTypesSelect struct {
	CreatedAt        string `json:"created_at"`
	CreditCostPerDay int32  `json:"credit_cost_per_day"`
	Id               string `json:"id"`
	Name             string `json:"name"`
}

type PublicEquipmentTypesInsert struct {
	CreatedAt        *string `json:"created_at"`
	CreditCostPerDay int32   `json:"credit_cost_per_day"`
	Id               *string `json:"id"`
	Name             string  `json:"name"`
}

type PublicEquipmentTypesUpdate struct {
	CreatedAt        *string `json:"created_at"`
	CreditCostPerDay *int32  `json:"credit_cost_per_day"`
	Id               *string `json:"id"`
	Name             *string `json:"name"`
}

type PublicEquipmentSelect struct {
	CreatedAt   string  `json:"created_at"`
	Description *string `json:"description"`
	Id          string  `json:"id"`
	ImagePath   *string `json:"image_path"`
	InternalId  string  `json:"internal_id"`
	IsArchived  bool    `json:"is_archived"`
	Name        *string `json:"name"`
	Status      string  `json:"status"`
	TypeId      string  `json:"type_id"`
	UpdatedAt   *string `json:"updated_at"`
}

type PublicEquipmentInsert struct {
	CreatedAt   *string `json:"created_at"`
	Description *string `json:"description"`
	Id          *string `json:"id"`
	ImagePath   *string `json:"image_path"`
	InternalId  string  `json:"internal_id"`
	IsArchived  *bool   `json:"is_archived"`
	Name        *string `json:"name"`
	Status      *string `json:"status"`
	TypeId      string  `json:"type_id"`
	UpdatedAt   *string `json:"updated_at"`
}

type PublicEquipmentUpdate struct {
	CreatedAt   *string `json:"created_at"`
	Description *string `json:"description"`
	Id          *string `json:"id"`
	ImagePath   *string `json:"image_path"`
	InternalId  *string `json:"internal_id"`
	IsArchived  *bool   `json:"is_archived"`
	Name        *string `json:"name"`
	Status      *string `json:"status"`
	TypeId      *string `json:"type_id"`
	UpdatedAt   *string `json:"updated_at"`
}

type PublicReservationsSelect struct {
	CreatedAt   string  `json:"created_at"`
	EndDate     string  `json:"end_date"`
	EquipmentId string  `json:"equipment_id"`
	Id          string  `json:"id"`
	StartDate   string  `json:"start_date"`
	Status      string  `json:"status"`
	UpdatedAt   *string `json:"updated_at"`
	UserId      string  `json:"user_id"`
}

type PublicReservationsInsert struct {
	CreatedAt   *string `json:"created_at"`
	EndDate     string  `json:"end_date"`
	EquipmentId string  `json:"equipment_id"`
	Id          *string `json:"id"`
	StartDate   string  `json:"start_date"`
	Status      *string `json:"status"`
	UpdatedAt   *string `json:"updated_at"`
	UserId      string  `json:"user_id"`
}

type PublicReservationsUpdate struct {
	CreatedAt   *string `json:"created_at"`
	EndDate     *string `json:"end_date"`
	EquipmentId *string `json:"equipment_id"`
	Id          *string `json:"id"`
	StartDate   *string `json:"start_date"`
	Status      *string `json:"status"`
	UpdatedAt   *string `json:"updated_at"`
	UserId      *string `json:"user_id"`
}

type PublicCreditHistorySelect struct {
	AdminId       *string `json:"admin_id"`
	Amount        int32   `json:"amount"`
	CreatedAt     string  `json:"created_at"`
	Description   *string `json:"description"`
	Id            string  `json:"id"`
	Reason        string  `json:"reason"`
	ReservationId *string `json:"reservation_id"`
	UserId        string  `json:"user_id"`
}

type PublicCreditHistoryInsert struct {
	AdminId       *string `json:"admin_id"`
	Amount        int32   `json:"amount"`
	CreatedAt     *string `json:"created_at"`
	Description   *string `json:"description"`
	Id            *string `json:"id"`
	Reason        string  `json:"reason"`
	ReservationId *string `json:"reservation_id"`
	UserId        string  `json:"user_id"`
}

type PublicCreditHistoryUpdate struct {
	AdminId       *string `json:"admin_id"`
	Amount        *int32  `json:"amount"`
	CreatedAt     *string `json:"created_at"`
	Description   *string `json:"description"`
	Id            *string `json:"id"`
	Reason        *string `json:"reason"`
	ReservationId *string `json:"reservation_id"`
	UserId        *string `json:"user_id"`
}

type PublicCreditRequestsSelect struct {
	AdminId     *string `json:"admin_id"`
	AdminNote   *string `json:"admin_note"`
	Amount      int32   `json:"amount"`
	CreatedAt   string  `json:"created_at"`
	Description string  `json:"description"`
	Id          string  `json:"id"`
	Status      string  `json:"status"`
	UpdatedAt   *string `json:"updated_at"`
	UserId      string  `json:"user_id"`
}

type PublicCreditRequestsInsert struct {
	AdminId     *string `json:"admin_id"`
	AdminNote   *string `json:"admin_note"`
	Amount      int32   `json:"amount"`
	CreatedAt   *string `json:"created_at"`
	Description string  `json:"description"`
	Id          *string `json:"id"`
	Status      *string `json:"status"`
	UpdatedAt   *string `json:"updated_at"`
	UserId      string  `json:"user_id"`
}

type PublicCreditRequestsUpdate struct {
	AdminId     *string `json:"admin_id"`
	AdminNote   *string `json:"admin_note"`
	Amount      *int32  `json:"amount"`
	CreatedAt   *string `json:"created_at"`
	Description *string `json:"description"`
	Id          *string `json:"id"`
	Status      *string `json:"status"`
	UpdatedAt   *string `json:"updated_at"`
	UserId      *string `json:"user_id"`
}

type PublicMaintenanceLogsSelect struct {
	AdminId        *string `json:"admin_id"`
	CreatedAt      string  `json:"created_at"`
	EquipmentId    string  `json:"equipment_id"`
	Id             string  `json:"id"`
	NewStatus      string  `json:"new_status"`
	Notes          *string `json:"notes"`
	PreviousStatus *string `json:"previous_status"`
}

type PublicMaintenanceLogsInsert struct {
	AdminId        *string `json:"admin_id"`
	CreatedAt      *string `json:"created_at"`
	EquipmentId    string  `json:"equipment_id"`
	Id             *string `json:"id"`
	NewStatus      string  `json:"new_status"`
	Notes          *string `json:"notes"`
	PreviousStatus *string `json:"previous_status"`
}

type PublicMaintenanceLogsUpdate struct {
	AdminId        *string `json:"admin_id"`
	CreatedAt      *string `json:"created_at"`
	EquipmentId    *string `json:"equipment_id"`
	Id             *string `json:"id"`
	NewStatus      *string `json:"new_status"`
	Notes          *string `json:"notes"`
	PreviousStatus *string `json:"previous_status"`
}

type PublicReservationHistorySelect struct {
	ChangedByUserId *string `json:"changed_by_user_id"`
	CreatedAt       string  `json:"created_at"`
	EndDate         string  `json:"end_date"`
	EquipmentId     string  `json:"equipment_id"`
	Id              string  `json:"id"`
	ReservationId   string  `json:"reservation_id"`
	StartDate       string  `json:"start_date"`
	Status          string  `json:"status"`
	UserId          string  `json:"user_id"`
}

type PublicReservationHistoryInsert struct {
	ChangedByUserId *string `json:"changed_by_user_id"`
	CreatedAt       *string `json:"created_at"`
	EndDate         string  `json:"end_date"`
	EquipmentId     string  `json:"equipment_id"`
	Id              *string `json:"id"`
	ReservationId   string  `json:"reservation_id"`
	StartDate       string  `json:"start_date"`
	Status          string  `json:"status"`
	UserId          string  `json:"user_id"`
}

type PublicReservationHistoryUpdate struct {
	ChangedByUserId *string `json:"changed_by_user_id"`
	CreatedAt       *string `json:"created_at"`
	EndDate         *string `json:"end_date"`
	EquipmentId     *string `json:"equipment_id"`
	Id              *string `json:"id"`
	ReservationId   *string `json:"reservation_id"`
	StartDate       *string `json:"start_date"`
	Status          *string `json:"status"`
	UserId          *string `json:"user_id"`
}

type PublicAnalyticsEquipmentStatsSelect struct {
	EquipmentId       *string  `json:"equipment_id"`
	EquipmentName     *string  `json:"equipment_name"`
	TotalDaysRented   *int64   `json:"total_days_rented"`
	TotalReservations *int64   `json:"total_reservations"`
	UtilizationRate   *float64 `json:"utilization_rate"`
}

type PublicAnalyticsUserStatsSelect struct {
	LastReservationDate *string `json:"last_reservation_date"`
	TotalCreditsSpent   *int64  `json:"total_credits_spent"`
	TotalReservations   *int64  `json:"total_reservations"`
	UserId              *string `json:"user_id"`
	Username            *string `json:"username"`
}

File: frontend/src/db/database.types.ts
export type Json =
  | string
  | number
  | boolean
  | null
  | { [key: string]: Json | undefined }
  | Json[];

export type Database = {
  // Allows to automatically instantiate createClient with right options
  // instead of createClient<Database, { PostgrestVersion: 'XX' }>(URL, KEY)
  __InternalSupabase: {
    PostgrestVersion: "13.0.5";
  };
  public: {
    Tables: {
      credit_history: {
        Row: {
          admin_id: string | null;
          amount: number;
          created_at: string;
          description: string | null;
          id: string;
          reason: Database["public"]["Enums"]["credit_transaction_reason"];
          reservation_id: string | null;
          user_id: string;
        };
        Insert: {
          admin_id?: string | null;
          amount: number;
          created_at?: string;
          description?: string | null;
          id?: string;
          reason: Database["public"]["Enums"]["credit_transaction_reason"];
          reservation_id?: string | null;
          user_id: string;
        };
        Update: {
          admin_id?: string | null;
          amount?: number;
          created_at?: string;
          description?: string | null;
          id?: string;
          reason?: Database["public"]["Enums"]["credit_transaction_reason"];
          reservation_id?: string | null;
          user_id?: string;
        };
        Relationships: [
          {
            foreignKeyName: "credit_history_admin_id_fkey";
            columns: ["admin_id"];
            isOneToOne: false;
            referencedRelation: "analytics_user_stats";
            referencedColumns: ["user_id"];
          },
          {
            foreignKeyName: "credit_history_admin_id_fkey";
            columns: ["admin_id"];
            isOneToOne: false;
            referencedRelation: "profiles";
            referencedColumns: ["id"];
          },
          {
            foreignKeyName: "credit_history_reservation_id_fkey";
            columns: ["reservation_id"];
            isOneToOne: false;
            referencedRelation: "reservations";
            referencedColumns: ["id"];
          },
          {
            foreignKeyName: "credit_history_user_id_fkey";
            columns: ["user_id"];
            isOneToOne: false;
            referencedRelation: "analytics_user_stats";
            referencedColumns: ["user_id"];
          },
          {
            foreignKeyName: "credit_history_user_id_fkey";
            columns: ["user_id"];
            isOneToOne: false;
            referencedRelation: "profiles";
            referencedColumns: ["id"];
          },
        ];
      };
      credit_requests: {
        Row: {
          admin_id: string | null;
          admin_note: string | null;
          amount: number;
          created_at: string;
          description: string;
          id: string;
          status: Database["public"]["Enums"]["credit_request_status"];
          updated_at: string | null;
          user_id: string;
        };
        Insert: {
          admin_id?: string | null;
          admin_note?: string | null;
          amount: number;
          created_at?: string;
          description: string;
          id?: string;
          status?: Database["public"]["Enums"]["credit_request_status"];
          updated_at?: string | null;
          user_id: string;
        };
        Update: {
          admin_id?: string | null;
          admin_note?: string | null;
          amount?: number;
          created_at?: string;
          description?: string;
          id?: string;
          status?: Database["public"]["Enums"]["credit_request_status"];
          updated_at?: string | null;
          user_id?: string;
        };
        Relationships: [
          {
            foreignKeyName: "credit_requests_admin_id_fkey";
            columns: ["admin_id"];
            isOneToOne: false;
            referencedRelation: "analytics_user_stats";
            referencedColumns: ["user_id"];
          },
          {
            foreignKeyName: "credit_requests_admin_id_fkey";
            columns: ["admin_id"];
            isOneToOne: false;
            referencedRelation: "profiles";
            referencedColumns: ["id"];
          },
          {
            foreignKeyName: "credit_requests_user_id_fkey";
            columns: ["user_id"];
            isOneToOne: false;
            referencedRelation: "analytics_user_stats";
            referencedColumns: ["user_id"];
          },
          {
            foreignKeyName: "credit_requests_user_id_fkey";
            columns: ["user_id"];
            isOneToOne: false;
            referencedRelation: "profiles";
            referencedColumns: ["id"];
          },
        ];
      };
      equipment: {
        Row: {
          created_at: string;
          description: string | null;
          id: string;
          image_path: string | null;
          internal_id: string;
          is_archived: boolean;
          name: string | null;
          status: Database["public"]["Enums"]["equipment_status"];
          type_id: string;
          updated_at: string | null;
        };
        Insert: {
          created_at?: string;
          description?: string | null;
          id?: string;
          image_path?: string | null;
          internal_id: string;
          is_archived?: boolean;
          name?: string | null;
          status?: Database["public"]["Enums"]["equipment_status"];
          type_id: string;
          updated_at?: string | null;
        };
        Update: {
          created_at?: string;
          description?: string | null;
          id?: string;
          image_path?: string | null;
          internal_id?: string;
          is_archived?: boolean;
          name?: string | null;
          status?: Database["public"]["Enums"]["equipment_status"];
          type_id?: string;
          updated_at?: string | null;
        };
        Relationships: [
          {
            foreignKeyName: "equipment_type_id_fkey";
            columns: ["type_id"];
            isOneToOne: false;
            referencedRelation: "equipment_types";
            referencedColumns: ["id"];
          },
        ];
      };
      equipment_types: {
        Row: {
          created_at: string;
          credit_cost_per_day: number;
          id: string;
          name: string;
        };
        Insert: {
          created_at?: string;
          credit_cost_per_day: number;
          id?: string;
          name: string;
        };
        Update: {
          created_at?: string;
          credit_cost_per_day?: number;
          id?: string;
          name?: string;
        };
        Relationships: [];
      };
      maintenance_logs: {
        Row: {
          admin_id: string | null;
          created_at: string;
          equipment_id: string;
          id: string;
          new_status: Database["public"]["Enums"]["equipment_status"];
          notes: string | null;
          previous_status:
            | Database["public"]["Enums"]["equipment_status"]
            | null;
        };
        Insert: {
          admin_id?: string | null;
          created_at?: string;
          equipment_id: string;
          id?: string;
          new_status: Database["public"]["Enums"]["equipment_status"];
          notes?: string | null;
          previous_status?:
            | Database["public"]["Enums"]["equipment_status"]
            | null;
        };
        Update: {
          admin_id?: string | null;
          created_at?: string;
          equipment_id?: string;
          id?: string;
          new_status?: Database["public"]["Enums"]["equipment_status"];
          notes?: string | null;
          previous_status?:
            | Database["public"]["Enums"]["equipment_status"]
            | null;
        };
        Relationships: [
          {
            foreignKeyName: "maintenance_logs_admin_id_fkey";
            columns: ["admin_id"];
            isOneToOne: false;
            referencedRelation: "analytics_user_stats";
            referencedColumns: ["user_id"];
          },
          {
            foreignKeyName: "maintenance_logs_admin_id_fkey";
            columns: ["admin_id"];
            isOneToOne: false;
            referencedRelation: "profiles";
            referencedColumns: ["id"];
          },
          {
            foreignKeyName: "maintenance_logs_equipment_id_fkey";
            columns: ["equipment_id"];
            isOneToOne: false;
            referencedRelation: "analytics_equipment_stats";
            referencedColumns: ["equipment_id"];
          },
          {
            foreignKeyName: "maintenance_logs_equipment_id_fkey";
            columns: ["equipment_id"];
            isOneToOne: false;
            referencedRelation: "equipment";
            referencedColumns: ["id"];
          },
        ];
      };
      profiles: {
        Row: {
          created_at: string;
          credit_balance: number;
          email: string;
          id: string;
          role: Database["public"]["Enums"]["user_role"];
          updated_at: string | null;
          username: string;
        };
        Insert: {
          created_at?: string;
          credit_balance?: number;
          email: string;
          id: string;
          role?: Database["public"]["Enums"]["user_role"];
          updated_at?: string | null;
          username: string;
        };
        Update: {
          created_at?: string;
          credit_balance?: number;
          email?: string;
          id?: string;
          role?: Database["public"]["Enums"]["user_role"];
          updated_at?: string | null;
          username?: string;
        };
        Relationships: [];
      };
      reservation_history: {
        Row: {
          changed_by_user_id: string | null;
          created_at: string;
          end_date: string;
          equipment_id: string;
          id: string;
          reservation_id: string;
          start_date: string;
          status: Database["public"]["Enums"]["reservation_status"];
          user_id: string;
        };
        Insert: {
          changed_by_user_id?: string | null;
          created_at?: string;
          end_date: string;
          equipment_id: string;
          id?: string;
          reservation_id: string;
          start_date: string;
          status: Database["public"]["Enums"]["reservation_status"];
          user_id: string;
        };
        Update: {
          changed_by_user_id?: string | null;
          created_at?: string;
          end_date?: string;
          equipment_id?: string;
          id?: string;
          reservation_id?: string;
          start_date?: string;
          status?: Database["public"]["Enums"]["reservation_status"];
          user_id?: string;
        };
        Relationships: [
          {
            foreignKeyName: "reservation_history_changed_by_user_id_fkey";
            columns: ["changed_by_user_id"];
            isOneToOne: false;
            referencedRelation: "analytics_user_stats";
            referencedColumns: ["user_id"];
          },
          {
            foreignKeyName: "reservation_history_changed_by_user_id_fkey";
            columns: ["changed_by_user_id"];
            isOneToOne: false;
            referencedRelation: "profiles";
            referencedColumns: ["id"];
          },
          {
            foreignKeyName: "reservation_history_equipment_id_fkey";
            columns: ["equipment_id"];
            isOneToOne: false;
            referencedRelation: "analytics_equipment_stats";
            referencedColumns: ["equipment_id"];
          },
          {
            foreignKeyName: "reservation_history_equipment_id_fkey";
            columns: ["equipment_id"];
            isOneToOne: false;
            referencedRelation: "equipment";
            referencedColumns: ["id"];
          },
          {
            foreignKeyName: "reservation_history_reservation_id_fkey";
            columns: ["reservation_id"];
            isOneToOne: false;
            referencedRelation: "reservations";
            referencedColumns: ["id"];
          },
          {
            foreignKeyName: "reservation_history_user_id_fkey";
            columns: ["user_id"];
            isOneToOne: false;
            referencedRelation: "analytics_user_stats";
            referencedColumns: ["user_id"];
          },
          {
            foreignKeyName: "reservation_history_user_id_fkey";
            columns: ["user_id"];
            isOneToOne: false;
            referencedRelation: "profiles";
            referencedColumns: ["id"];
          },
        ];
      };
      reservations: {
        Row: {
          created_at: string;
          end_date: string;
          equipment_id: string;
          id: string;
          start_date: string;
          status: Database["public"]["Enums"]["reservation_status"];
          updated_at: string | null;
          user_id: string;
        };
        Insert: {
          created_at?: string;
          end_date: string;
          equipment_id: string;
          id?: string;
          start_date: string;
          status?: Database["public"]["Enums"]["reservation_status"];
          updated_at?: string | null;
          user_id: string;
        };
        Update: {
          created_at?: string;
          end_date?: string;
          equipment_id?: string;
          id?: string;
          start_date?: string;
          status?: Database["public"]["Enums"]["reservation_status"];
          updated_at?: string | null;
          user_id?: string;
        };
        Relationships: [
          {
            foreignKeyName: "reservations_equipment_id_fkey";
            columns: ["equipment_id"];
            isOneToOne: false;
            referencedRelation: "analytics_equipment_stats";
            referencedColumns: ["equipment_id"];
          },
          {
            foreignKeyName: "reservations_equipment_id_fkey";
            columns: ["equipment_id"];
            isOneToOne: false;
            referencedRelation: "equipment";
            referencedColumns: ["id"];
          },
          {
            foreignKeyName: "reservations_user_id_fkey";
            columns: ["user_id"];
            isOneToOne: false;
            referencedRelation: "analytics_user_stats";
            referencedColumns: ["user_id"];
          },
          {
            foreignKeyName: "reservations_user_id_fkey";
            columns: ["user_id"];
            isOneToOne: false;
            referencedRelation: "profiles";
            referencedColumns: ["id"];
          },
        ];
      };
    };
    Views: {
      analytics_equipment_stats: {
        Row: {
          equipment_id: string | null;
          equipment_name: string | null;
          total_days_rented: number | null;
          total_reservations: number | null;
          utilization_rate: number | null;
        };
        Relationships: [];
      };
      analytics_user_stats: {
        Row: {
          last_reservation_date: string | null;
          total_credits_spent: number | null;
          total_reservations: number | null;
          user_id: string | null;
          username: string | null;
        };
        Relationships: [];
      };
    };
    Functions: {
      [_ in never]: never;
    };
    Enums: {
      credit_request_status: "PENDING" | "APPROVED" | "DENIED";
      credit_transaction_reason:
        | "reservation_charge"
        | "reservation_refund"
        | "reservation_adjustment"
        | "admin_adjustment"
        | "work_credit";
      equipment_status: "ok" | "broken" | "blocked";
      reservation_status: "PENDING" | "RENTED" | "RETURNED" | "DENIED";
      user_role: "user" | "admin" | "super_admin";
    };
    CompositeTypes: {
      [_ in never]: never;
    };
  };
};

type DatabaseWithoutInternals = Omit<Database, "__InternalSupabase">;

type DefaultSchema = DatabaseWithoutInternals[Extract<
  keyof Database,
  "public"
>];

export type Tables<
  DefaultSchemaTableNameOrOptions extends
    | keyof (DefaultSchema["Tables"] & DefaultSchema["Views"])
    | { schema: keyof DatabaseWithoutInternals },
  TableName extends DefaultSchemaTableNameOrOptions extends {
    schema: keyof DatabaseWithoutInternals;
  }
    ? keyof (DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Tables"] &
        DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Views"])
    : never = never,
> = DefaultSchemaTableNameOrOptions extends {
  schema: keyof DatabaseWithoutInternals;
}
  ? (DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Tables"] &
      DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Views"])[TableName] extends {
      Row: infer R;
    }
    ? R
    : never
  : DefaultSchemaTableNameOrOptions extends keyof (DefaultSchema["Tables"] &
        DefaultSchema["Views"])
    ? (DefaultSchema["Tables"] &
        DefaultSchema["Views"])[DefaultSchemaTableNameOrOptions] extends {
        Row: infer R;
      }
      ? R
      : never
    : never;

export type TablesInsert<
  DefaultSchemaTableNameOrOptions extends
    | keyof DefaultSchema["Tables"]
    | { schema: keyof DatabaseWithoutInternals },
  TableName extends DefaultSchemaTableNameOrOptions extends {
    schema: keyof DatabaseWithoutInternals;
  }
    ? keyof DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Tables"]
    : never = never,
> = DefaultSchemaTableNameOrOptions extends {
  schema: keyof DatabaseWithoutInternals;
}
  ? DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Tables"][TableName] extends {
      Insert: infer I;
    }
    ? I
    : never
  : DefaultSchemaTableNameOrOptions extends keyof DefaultSchema["Tables"]
    ? DefaultSchema["Tables"][DefaultSchemaTableNameOrOptions] extends {
        Insert: infer I;
      }
      ? I
      : never
    : never;

export type TablesUpdate<
  DefaultSchemaTableNameOrOptions extends
    | keyof DefaultSchema["Tables"]
    | { schema: keyof DatabaseWithoutInternals },
  TableName extends DefaultSchemaTableNameOrOptions extends {
    schema: keyof DatabaseWithoutInternals;
  }
    ? keyof DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Tables"]
    : never = never,
> = DefaultSchemaTableNameOrOptions extends {
  schema: keyof DatabaseWithoutInternals;
}
  ? DatabaseWithoutInternals[DefaultSchemaTableNameOrOptions["schema"]]["Tables"][TableName] extends {
      Update: infer U;
    }
    ? U
    : never
  : DefaultSchemaTableNameOrOptions extends keyof DefaultSchema["Tables"]
    ? DefaultSchema["Tables"][DefaultSchemaTableNameOrOptions] extends {
        Update: infer U;
      }
      ? U
      : never
    : never;

export type Enums<
  DefaultSchemaEnumNameOrOptions extends
    | keyof DefaultSchema["Enums"]
    | { schema: keyof DatabaseWithoutInternals },
  EnumName extends DefaultSchemaEnumNameOrOptions extends {
    schema: keyof DatabaseWithoutInternals;
  }
    ? keyof DatabaseWithoutInternals[DefaultSchemaEnumNameOrOptions["schema"]]["Enums"]
    : never = never,
> = DefaultSchemaEnumNameOrOptions extends {
  schema: keyof DatabaseWithoutInternals;
}
  ? DatabaseWithoutInternals[DefaultSchemaEnumNameOrOptions["schema"]]["Enums"][EnumName]
  : DefaultSchemaEnumNameOrOptions extends keyof DefaultSchema["Enums"]
    ? DefaultSchema["Enums"][DefaultSchemaEnumNameOrOptions]
    : never;

export type CompositeTypes<
  PublicCompositeTypeNameOrOptions extends
    | keyof DefaultSchema["CompositeTypes"]
    | { schema: keyof DatabaseWithoutInternals },
  CompositeTypeName extends PublicCompositeTypeNameOrOptions extends {
    schema: keyof DatabaseWithoutInternals;
  }
    ? keyof DatabaseWithoutInternals[PublicCompositeTypeNameOrOptions["schema"]]["CompositeTypes"]
    : never = never,
> = PublicCompositeTypeNameOrOptions extends {
  schema: keyof DatabaseWithoutInternals;
}
  ? DatabaseWithoutInternals[PublicCompositeTypeNameOrOptions["schema"]]["CompositeTypes"][CompositeTypeName]
  : PublicCompositeTypeNameOrOptions extends keyof DefaultSchema["CompositeTypes"]
    ? DefaultSchema["CompositeTypes"][PublicCompositeTypeNameOrOptions]
    : never;

export const Constants = {
  public: {
    Enums: {
      credit_request_status: ["PENDING", "APPROVED", "DENIED"],
      credit_transaction_reason: [
        "reservation_charge",
        "reservation_refund",
        "reservation_adjustment",
        "admin_adjustment",
        "work_credit",
      ],
      equipment_status: ["ok", "broken", "blocked"],
      reservation_status: ["PENDING", "RENTED", "RETURNED", "DENIED"],
      user_role: ["user", "admin", "super_admin"],
    },
  },
} as const;
